File: backend-node/src/routes/cards.js
import express from 'express';
import { pool } from '../db.js';
import { formatCardHolder, generateCardNumber, generateCvv, generateExpDate } from '../utils.js';

const router = express.Router();

async function createUniqueCard(client, userId, username) {
  for (let attempt = 0; attempt < 30; attempt += 1) {
    try {
      const result = await client.query(
        `INSERT INTO cards (user_id, card_number, card_holder, balance, exp_date, cvv)
         VALUES ($1, $2, $3, 0, $4, $5)
         RETURNING id, user_id, card_number, card_holder, balance, exp_date, cvv, created_at`,
        [userId, generateCardNumber(), formatCardHolder(username), generateExpDate(), generateCvv()]
      );
      return result.rows[0];
    } catch (error) {
      if (error.code !== '23505') throw error;
    }
  }
  throw new Error('Unable to generate a unique card number');
}

router.get('/', async (req, res, next) => {
  try {
    const [cards, user] = await Promise.all([
      pool.query(
        `SELECT id, user_id, card_number, card_holder, balance, exp_date, cvv, created_at
         FROM cards WHERE user_id = $1 ORDER BY created_at ASC`,
        [req.user.id]
      ),
      pool.query(`SELECT active_rating_card_id FROM users WHERE id = $1`, [req.user.id])
    ]);
    return res.json({ active_rating_card_id: user.rows[0]?.active_rating_card_id || null, cards: cards.rows });
  } catch (error) {
    return next(error);
  }
});

router.post('/create', async (req, res, next) => {
  const client = await pool.connect();
  try {
    await client.query('BEGIN');
    const userResult = await client.query(`SELECT username FROM users WHERE id = $1 FOR UPDATE`, [req.user.id]);
    if (!userResult.rows[0]) {
      await client.query('ROLLBACK');
      return res.status(404).json({ error: 'User not found' });
    }
    const count = await client.query(`SELECT COUNT(*)::int AS count FROM cards WHERE user_id = $1`, [req.user.id]);
    if (count.rows[0].count >= 5) {
      await client.query('ROLLBACK');
      return res.status(400).json({ error: 'Maximum of 5 cards per user' });
    }
    const card = await createUniqueCard(client, req.user.id, userResult.rows[0].username);
    await client.query('COMMIT');
    return res.status(201).json(card);
  } catch (error) {
    await client.query('ROLLBACK');
    return next(error);
  } finally {
    client.release();
  }
});

export default router;

File: backend-python/app/routes/cards.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import func, select
from sqlalchemy.orm import Session
from ..auth import get_current_user
from ..database import get_db
from ..models import Card, User
from ..services import create_unique_card

router = APIRouter(prefix='/api/cards', tags=['cards'], dependencies=[Depends(get_current_user)])

@router.get('')
def get_cards(user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    cards = list(db.scalars(select(Card).where(Card.user_id == user.id).order_by(Card.created_at.asc())))
    return {
        'active_rating_card_id': str(user.active_rating_card_id) if user.active_rating_card_id else None,
        'cards': [
            {'id': str(c.id), 'user_id': str(c.user_id), 'card_number': c.card_number, 'card_holder': c.card_holder, 'balance': str(c.balance), 'exp_date': c.exp_date, 'cvv': c.cvv, 'created_at': c.created_at.isoformat()}
            for c in cards
        ]
    }

@router.post('/create', status_code=201)
def create_card(user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    with db.begin():
        count = db.scalar(select(func.count(Card.id)).where(Card.user_id == user.id)) or 0
        if count >= 5:
            raise HTTPException(400, 'Maximum of 5 cards per user')
        db.refresh(user)
        card = create_unique_card(db, user.id, user.username)
    return {'id': str(card.id), 'user_id': str(card.user_id), 'card_number': card.card_number, 'card_holder': card.card_holder, 'balance': str(card.balance), 'exp_date': card.exp_date, 'cvv': card.cvv, 'created_at': card.created_at.isoformat()}
